src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,575 @@
+        self.num_replicas = 0
+        self._work_registry = {}
+
+        self._work_cls = work_cls
+        self._work_args = work_args
+        self._work_kwargs = work_kwargs
+
+        self._input_type = input_type
+        self._output_type = output_type
+        self.autoscale_interval = autoscale_interval
+        self.max_batch_size = max_batch_size
+
+        if max_replicas < min_replicas:
+            raise ValueError(
+                f"`max_replicas={max_replicas}` must be less than or equal to `min_replicas={min_replicas}`."
+            )
+        self.max_replicas = max_replicas
+        self.min_replicas = min_replicas
+        self._last_autoscale = time.time()
+        self.fake_trigger = 0
+
+        endpoint = endpoint or "api/predict"
+        self.load_balancer = _LoadBalancer(
+            input_type=self._input_type,
+            output_type=self._output_type,
+            endpoint=endpoint,
+            max_batch_size=max_batch_size,
+            timeout_batching=timeout_batching,
+            cache_calls=True,
+            parallel=True,
+        )
+        for _ in range(min_replicas):
+            work = self.create_work()
+            self.add_work(work)
+
+    @property
+    def workers(self) -> List[LightningWork]:
+        return [self.get_work(i) for i in range(self.num_replicas)]
+
+    def create_work(self) -> LightningWork:
+        """Replicates a LightningWork instance with args and kwargs provided via ``__init__``."""
+        # TODO: Remove `start_with_flow=False` for faster initialization on the cloud
+        return self._work_cls(*self._work_args, **self._work_kwargs, start_with_flow=False)
+
+    def add_work(self, work) -> str:
+        """Adds a new LightningWork instance.
+
+        Returns:
+            The name of the new work attribute.
+        """
+        work_attribute = uuid.uuid4().hex
+        work_attribute = f"worker_{self.num_replicas}_{str(work_attribute)}"
+        setattr(self, work_attribute, work)
+        self._work_registry[self.num_replicas] = work_attribute
+        self.num_replicas += 1
+        return work_attribute
+
+    def remove_work(self, index: int) -> str:
+        """Removes the ``index`` th LightningWork instance."""
+        work_attribute = self._work_registry[index]
+        del self._work_registry[index]
+        work = getattr(self, work_attribute)
+        work.stop()
+        self.num_replicas -= 1
+        return work_attribute
+
+    def get_work(self, index: int) -> LightningWork:
+        """Returns the ``index`` th LightningWork instance."""
+        work_attribute = self._work_registry[index]
+        work = getattr(self, work_attribute)
+        return work
+
+    def run(self):
+        if not self.load_balancer.is_running:
+            self.load_balancer.run()
+
+        for work in self.workers:
+            work.run()
+
+        if self.load_balancer.url:
+            self.fake_trigger += 1  # Note: change state to keep calling `run`.
```

Comment:
Let's swap it to use `self.schedule` eventually, will need a bit of testing to check it works correctly so let's keep it for a follow-up